x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/ApiKeyFieldNameTranslators.java

```diff
@@ -38,37 +61,243 @@ public class ApiKeyFieldNameTranslators {
+     * The optional {@param visitor} can be used to collect all the translated field names.
+     */
+    public static void translateFieldSortBuilders(
+        List<FieldSortBuilder> fieldSortBuilders,
+        SearchSourceBuilder searchSourceBuilder,
+        @Nullable Consumer<String> visitor
+    ) {
+        final Consumer<String> fieldNameVisitor = visitor != null ? visitor : ignored -> {};
+        fieldSortBuilders.forEach(fieldSortBuilder -> {
+            if (fieldSortBuilder.getNestedSort() != null) {
+                throw new IllegalArgumentException("nested sorting is not supported for API Key query");
+            }
+            if (FieldSortBuilder.DOC_FIELD_NAME.equals(fieldSortBuilder.getFieldName())) {
+                searchSourceBuilder.sort(fieldSortBuilder);
+            } else {
+                final String translatedFieldName = translate(fieldSortBuilder.getFieldName());
+                fieldNameVisitor.accept(translatedFieldName);
+                if (translatedFieldName.equals(fieldSortBuilder.getFieldName())) {
+                    searchSourceBuilder.sort(fieldSortBuilder);
+                } else {
+                    final FieldSortBuilder translatedFieldSortBuilder = new FieldSortBuilder(translatedFieldName).order(
+                        fieldSortBuilder.order()
+                    )
+                        .missing(fieldSortBuilder.missing())
+                        .unmappedType(fieldSortBuilder.unmappedType())
+                        .setFormat(fieldSortBuilder.getFormat());
+
+                    if (fieldSortBuilder.sortMode() != null) {
+                        translatedFieldSortBuilder.sortMode(fieldSortBuilder.sortMode());
+                    }
+                    if (fieldSortBuilder.getNestedSort() != null) {
+                        translatedFieldSortBuilder.setNestedSort(fieldSortBuilder.getNestedSort());
+                    }
+                    if (fieldSortBuilder.getNumericType() != null) {
+                        translatedFieldSortBuilder.setNumericType(fieldSortBuilder.getNumericType());
+                    }
+                    searchSourceBuilder.sort(translatedFieldSortBuilder);
+                }
+            }
+        });
+    }
+
+    /**
+     * Deep copies the passed-in {@param queryBuilder} translating all the field names, from query level to index level,
+     * see {@link  #translate}. In general, the returned builder should create the same query as if the query were
+     * created by the passed in {@param queryBuilder}, only with the field names translated.
+     * Field name patterns (including "*"), are also replaced with the explicit index level field names whose
+     * associated query level field names match the pattern.
+     * The optional {@param visitor} can be used to collect all the translated field names.
+     */
+    public static QueryBuilder translateQueryBuilderFields(QueryBuilder queryBuilder, @Nullable Consumer<String> visitor) {
```

Comment:
This is the end goal of this PR.
I needed to reuse this method in the aggs builder (for processing the queries of filter(s) aggs).
As part of increasing the visibility of this method, I also wanted to ensure that we return a deep copy of the passed-in query builder. This is not a strict requirement, but it's a more appropriately clean interface given the increased visibility.
Originally, we copied most builder instance types, and modified in-place some of them. The notable example is that of the `SimpleQueryStringBuilder` which is onerous to copy.